### PR TITLE
Add first version of operator boilerplate

### PR DESCRIPTION
**Pull request by tuommaki**
First version of cluster-operator boilerplate code. Implementation is
based on microkit and operatorkit implementations and closely follows
high level architecture of azure-operator and endpoint-operator.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+# How to contribute
```
Comment (xh3b4sd) [inline CONTRIBUTING.md]: The legal magic could be a separate PR.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,7 @@
+FROM alpine:3.5
+
+RUN apk add --no-cache ca-certificates
+
+ADD ./cluster-operator /cluster-operator
+
+ENTRYPOINT ["/cluster-operator"]
```
Comment (xh3b4sd) [inline Dockerfile]: The Docker magic could be a separate PR.

File: helm/endpoint-operator-chart/Chart.yaml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+name: cluster-operator-chart
+version: 0.0.1-[[ .SHA ]]
```
Comment (xh3b4sd) [inline helm/endpoint-operator-chart/Chart.yaml]: The Helm magic could be separated as well.

File: flag/service/kubernetes/tls/tls.go [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package tls
+
+// TLS is a data structure for Kubernetes TLS configuration with command line
+// flags
+type TLS struct {
+	CAFile  string
+	CrtFile string
+	KeyFile string
+}
```
Comment (xh3b4sd) [inline flag/service/kubernetes/tls/tls.go]: The actual go project structure could have its separate PR.

File: main.go [hunk 5]
```diff
@@ -1,12 +1,149 @@
...
+			serviceConfig.GitCommit = gitCommit
+			serviceConfig.Name = name
+			serviceConfig.Source = source
+
+			newService, err = service.New(serviceConfig)
+			if err != nil {
+				panic(fmt.Sprintf("%#v\n", microerror.Maskf(err, "service.New")))
+			}
+
+			// TODO: Comment the lifecycle of this goroutine
```
Comment (xh3b4sd) [inline main.go]: Here is a TODO left. I think such docs would be good to have at the `Boot` method. Then it is visible in the godoc and actually useful for users.

File: service/operator/service.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
+package operator
```
Comment (xh3b4sd) [inline service/operator/service.go]: I think this is not necessary anymore. There are probably still operators who have this dedicated service but this is rather legacy. Check the kvm-operator for instance. It does not have this anymore.

File: service/version_bundles.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
+package service
+
+import (
+	"time"
+
+	"github.com/giantswarm/versionbundle"
+)
+
+func newVersionBundles() []versionbundle.Bundle {
```
Comment (xh3b4sd) [inline service/version_bundles.go]: I think to begin with we do not need version bundles in this operator. It does not implement the business logic the version bundles suggest. This is already done by the other operators. An operator only needs version bundles to guarantee versioning and communication with the end user. The cluster-operator is an internal detail we can more easily change however we like.

File: Gopkg.toml [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  "k8s.io/api/core/v1",
+]
+
+[[constraint]]
+  name = "k8s.io/api"
+  version = "kubernetes-1.8.5"
+
+[[constraint]]
+  name = "github.com/giantswarm/apiextensions"
+  branch = "master"
```
Comment (xh3b4sd) [inline Gopkg.toml]: Super pedantic nitpick. The order is alphabetical in other files.

File: Gopkg.toml [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+#  name = "github.com/x/y"
+#  version = "2.4.0"
+
+required = [
+  "k8s.io/api/core/v1",
+]
+
+[[constraint]]
+  name = "k8s.io/api"
+  version = "kubernetes-1.8.5"
```
Comment (xh3b4sd) [inline Gopkg.toml]: We said we want to import all the necessary deps and pin them because dep does not inherit deps like glide did. Check kvm-operator for instance. This might not yet be a problem but I can see this being an issue as soon as you start working with the operator in an installation.

**Comment by xh3b4sd** [comment]
Note that the CI fails.

File: helm/endpoint-operator-chart/Chart.yaml (hunk 3 above)
Comment (kopiczko) [reply helm/endpoint-operator-chart/Chart.yaml]: I also think this PR is too big. I'd separate helm, service stuff. I'd add minimal set of dependencies (no kubernetes deps). vendor dir will change anyway when you add importing code.

File: Gopkg.toml (hunk 9 above)
Comment (tuommaki) [reply Gopkg.toml]: There's probably something I don't understand, but do we want to pin the deps or not? Here: https://github.com/giantswarm/kvm-operator/blob/master/Gopkg.toml kvm-operator doesn't pin the versions. But the above referenced piece does. I added it after getting compilation errors and search in Slack brought a discussion which stated that said piece should be used (I think it was Pawel mentioning this ~2 weeks ago).

File: flag/service/kubernetes/tls/tls.go (hunk 4 above)
Comment (tuommaki) [reply flag/service/kubernetes/tls/tls.go]: When submitting the Go project structure, shall I include or exclude `dep` files (`Gopkg.*`)?

File: Gopkg.toml (hunk 9 above)
Comment (xh3b4sd) [reply Gopkg.toml]: Oh wow. I see. This is odd. I had lots of trouble with cluster-service and though kvm-operator is sorted already. See https://github.com/giantswarm/cluster-service/blob/master/Gopkg.toml. I found it was safer for updating dependencies using dep when you are explicit. The transition was just not possible without the pins in the cluster service so there was a conclusion also announced in #news-dev that we should always be as explicit as possible.

File: flag/service/kubernetes/tls/tls.go (hunk 4 above)
Comment (xh3b4sd) [reply flag/service/kubernetes/tls/tls.go]: The dep files have always have to be there, otherwise the code will not build.

File: service/version_bundles.go (hunk 7 above)
Comment (tuommaki) [reply service/version_bundles.go]: As this is not needed, there's no HTTP interface in the cluster-operator (unless we want to provide healthz for k8s, but even still it would be easy to implement with stdlib) so I'm wondering if microkit is needed for anything? I grasp that operatorkit is partially build with microkit in mind, but it's not a hard dependency there - right? So what do you think? Is there a reason for microkit usage here and what would be the benefits?

Comment (xh3b4sd) [reply service/version_bundles.go]: The core process is managed via microkit. It is the standard for running our daemons. The command line interface is the same for all services, except the flags it consumes at runtime. The whole flag and config file management is based on microkit. The microendpoints integrate directly with it and we want that for the health checks.

**Comment by tuommaki** [comment]
Ok. In a moment I'll submit the split PRs. One question still: Shall I include or exclude the `vendor` directory from the initial Go project structure PR? I think that populated vendor directories make PRs hard to follow, but I understand why we do it. Still, should I include it in the first PR or just make a separate one for that one too once the Go project gets approval?

**Comment by kopiczko** [comment]
W/o vendor it won't build. I have 0 issue reviewing PR with vendor directory. Hardly ever you have another directory starting with letter after `v` in the alphabet.
If possible make sure to push minimal set of dependencies and add them as needed.

**Comment by xh3b4sd** [comment]
Please make sure that you delete branches that are merged or closed. If necessary we can restore them but having all the clutter over time makes management only more complicated in a repo.
